File: Java/src/covariance/abs/AbsVertex.java
package covariance.abs;

import java.util.HashSet;

public abstract class AbsVertex
{
    private Integer label;
    private HashSet<AbsEdge> incidentEdges;
    private AbsGraph graph;

    public AbsVertex(Integer label)
    {
        this.label = label;
        incidentEdges = new HashSet<AbsEdge>();
    }
    
    public void unbindGraph()
    {
        for (AbsEdge e : incidentEdges)
            graph.removeEdge(e);

        this.graph = null;
    }

    public AbsGraph getGraph()
    {
        return graph;
    }
    
    public void setGraph(AbsGraph g)
    {
        if (graph != null) unbindGraph();
        this.graph = g;
    }

    public HashSet<AbsEdge> getIncidentEdges()
    {
        return incidentEdges;
    }

    public String toString()
    {
        return Integer.toString(label);
    }
}

File: scala/src/covariance/abs/AbsVertex.scala
package covariance.abs

import scala.collection.mutable.HashSet

abstract class AbsVertex(private var label:Integer) {
  
  def unbindGraph(): Unit = {
    incidentEdges.foreach(e => graph.removeEdge(e));
    this.graph = null
  }

  def getGraph: AbsGraph = graph

  def setGraph(g: AbsGraph): Unit = {
    if (graph != null)
      unbindGraph()

    this.graph = g
  }

  def getIncidentEdges: HashSet[AbsEdge] = incidentEdges

  override def toString(): String = label.toString()
  
  private var incidentEdges: HashSet[AbsEdge] = new HashSet[AbsEdge]
  private var graph: AbsGraph = null
}
